### PR TITLE
EnJso (Garo) OK and documented

### PR DESCRIPTION
**Pull request by tom-overton**
Some of this documentation is a little sparse, but I'm hoping that Garo Master may shine a light on some of the odder aspects of this actor.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 1]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
+ * appear in one of four different ways, activating the ring of fire, etc. When the cutscene is over, the Garo will
+ * jump back to start the fight.
+ */
+void EnJso_HandleIntroCutscene(EnJso* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
+    f32 curFrame = this->skelAnime.curFrame;
+    s16 showTextbox;
+    s16 diffToTargetRobeRightRot;
+
+    switch (this->introCsState) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:361]: Could you add a default to this case and maintain matching? (sometimes these

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 2]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
+ * Unlike most enemies, this will make the Garo play its damaged animation rather than stopping all animations.
+ */
+void EnJso_SetupStunned(EnJso* this) {
+    Vec3f knockbackVelocity;
+
+    AudioSfx_SetChannelIO(&this->actor.projectedPos, NA_SE_EN_ANSATSUSYA_DASH_2, 0);
+    EnJso_ChangeAnim(this, EN_JSO_ANIM_DAMAGED);
+
+    //! @bug: This assignment is immediately overriden below.
+    this->timer = 30;
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:982]: While this is a little jank, I'm not sure this is really a bug since it doesn't affect the behaviour in any way?

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 3]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
+/**
+ * Handles the first part of the dialogue the Garo gives when defeated.
+ */
+void EnJso_FallDownAndTalk(EnJso* this, PlayState* play) {
+    f32 curFrame = this->skelAnime.curFrame;
+    Player* player = GET_PLAYER(play);
+
+    if (!(curFrame < this->animEndFrame) && !Play_InCsMode(play) && (play->msgCtx.msgLength == 0)) {
+        if (fabsf(player->actor.world.pos.y - this->actor.world.pos.y) < 60.0f) {
+            Player* player2 = GET_PLAYER(play);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1166]: Do you think this `player2` could be a fake match? It appears a couple times in this actor, maybe it was just the style of the dev?

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 4]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
+    Vec3f swordTipPos;
+    Vec3f swordBasePos;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AF110.s")
+    Matrix_Push();
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AF28C.s")
+    if (limbIndex == GARO_LIMB_LEFT_SWORD) {
+        if (this->swordState == EN_JSO_SWORD_STATE_KNOCKED_OUT_OF_HANDS) {
+            Actor_SpawnBodyParts(&this->actor, play, 15, dList);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1557]:
```suggestion
            Actor_SpawnBodyParts(&this->actor, play, ENPART_TYPE_15, dList);
```
Might need to include `z_en_part.h` here

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 5]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
+            EffectBlure_AddVertex(Effect_GetByIndex(this->leftSwordBlureIndex), &swordTipPos, &swordBasePos);
+        } else if (this->disableBlure == true) {
+            EffectBlure_AddSpace(Effect_GetByIndex(this->leftSwordBlureIndex));
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AF3C0.s")
+    if (limbIndex == GARO_LIMB_RIGHT_SWORD) {
+        if (this->swordState == EN_JSO_SWORD_STATE_KNOCKED_OUT_OF_HANDS) {
+            Actor_SpawnBodyParts(&this->actor, play, 15, dList);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1581]:
```suggestion
            Actor_SpawnBodyParts(&this->actor, play, ENPART_TYPE_15, dList);
```

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 6]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
+        Matrix_MultVec3f(&sSwordBaseQuadOffset, &this->leftSwordCollider.dim.quad[0]);
+        Collider_SetQuadVertices(&this->leftSwordCollider, &this->leftSwordCollider.dim.quad[0],
+                                 &this->leftSwordCollider.dim.quad[1], &this->leftSwordCollider.dim.quad[2],
+                                 &this->leftSwordCollider.dim.quad[3]);
+        Matrix_MultVec3f(&sSwordTipOffset, &swordTipPos);
+        Matrix_MultVec3f(&sSwordBaseOffset, &swordBasePos);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AF368.s")
+        if (((this->action == EN_JSO_ACTION_SLASH) || (this->action == EN_JSO_ACTION_DASH_ATTACK)) &&
+            (!this->disableBlure)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1572]:
```suggestion
            !this->disableBlure) {
```

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 7]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
+        Matrix_MultVec3f(&sSwordBaseQuadOffset, &this->rightSwordCollider.dim.quad[0]);
+        Collider_SetQuadVertices(&this->rightSwordCollider, &this->rightSwordCollider.dim.quad[0],
+                                 &this->rightSwordCollider.dim.quad[1], &this->rightSwordCollider.dim.quad[2],
+                                 &this->rightSwordCollider.dim.quad[3]);
+        Matrix_MultVec3f(&sSwordTipOffset, &swordTipPos);
+        Matrix_MultVec3f(&sSwordBaseOffset, &swordBasePos);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AF440.s")
+        if (((this->action == EN_JSO_ACTION_SLASH) || (this->action == EN_JSO_ACTION_DASH_ATTACK)) &&
+            (!this->disableBlure)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1597]:
```suggestion
            !this->disableBlure) {
```

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 8]
```diff
@@ -81,106 +158,1547 @@ ActorInit En_Jso_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AFAF4.s")
+    if (this->afterimageCount > 0) {
+        s32 i;
+        s32 index = this->afterimageIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AFC10.s")
+        for (i = 0; i < this->afterimageCount; i++) {
+            if (sAfterimageAlpha[i] == 0) {
+                continue;
+            } else {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1655]:
```suggestion
            if (sAfterimageAlpha[i] == 0) {
                continue;
            }
```
Since there's already a break in logic, I don't think you need the `else`?

**Review by engineer124 — COMMENTED** [review]
Really nice work!

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 1 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:361]: No, I tried to insert a default case in every possible place in this switch-block, and it always breaks matching. It's the only switch in this file that refuses to be okay with a default case.

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 3 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1166]: One thing we can say for certain is that calling `GET_PLAYER` a second time in this if-block is almost assuredly real. Here's the function you highlighted with the `player2` stuff entirely removed. Notice the missing `lw v0,0x1ccc(a3)` in the right column; there are two in the original, but only one in the current:
image

Anytime you see a `player2` in this file (which is only twice, mind you), the key thing to note is that the second `GET_PLAYER` macro is probably real, because you have to do it twice in the function to get that `lw v0,0x1ccc(a3)` to show up twice. Now, could you potentially match this by doing `GET_PLAYER` twice but only using the first `player`? Maybe, but I've never gotten it to match like that, and it doesn't feel worth it because we have to call `GET_PLAYER` twice anyways, so we might as well use it.

So in short, I think this is actually real and just a weird stylistic quirk. If you can figure out how to get that second load to show up with just a single `GET_PLAYER`, I'd be fully willing to look into this further. But from what I can tell, we need two `GET_PLAYER`s here.

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 4 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1557]: Huh, wondered about this one, because *none* of the calls to `Actor_SpawnBodyParts` are using this enum, so I figured it must not exist. Let me update them to all use the enum.

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 3 above)
Comment (engineer124) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1166]: Yeah, that seems like enough of an argument for me, thanks for looking into it

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 4 above)
Comment (engineer124) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1557]: Yeah, I noticed that too. I was going to do that in my next anim cleanup PR but it'd be great if you wanted to do that right now. A step further would be to also make a param setter macro like clear-tag, but I'll leave it up to you for how far you want to take

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 8 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1655]: At one point, I think this was written in a way where doing this broke matching, but not anymore. Thanks for bringing it up!

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 4 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1557]: I just made it like EnClearTag, however I did not bother to name the part types so they're still `ENPART_TYPE_15` for example. I figure they can be named in a follow-up PR; changing what other actors pass into `Actor_SpawnBodyParts` already expands the scope of this PR a bit, so let's not make it any bigger.

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 9]
```diff
@@ -81,106 +159,1546 @@ ActorInit En_Jso_InitVars = {
...
+
+            if (this->introCsTimer >= 10) {
+                this->swordState = EN_JSO_SWORD_STATE_BOTH_DRAWN;
+                Actor_PlaySfx(&this->actor, NA_SE_EN_ANSATSUSYA_SWORD);
+                if ((this->actor.parent != NULL) && (this->actor.parent->update != NULL)) {
+                    EnEncount3* parent = (EnEncount3*)this->actor.parent;
+
+                    if (parent->actor.update != NULL) {
+                        parent->unk148 = 1;
+                        play->unk_18880 = 1;
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:556]: Doesn't need to be figured out in this PR, but `play->unk_18880` does interest me. It only shows up with Garo and the Garo spawner so far, does it also appear in the Garo Master overlay?

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 10]
```diff
@@ -81,106 +159,1546 @@ ActorInit En_Jso_InitVars = {
...
+            EnJso_SetupKnockedBack(this);
+            return;
+        }
+    }
+
+    if ((this->actor.velocity.y < 0.0f) && (this->actor.bgCheckFlags & BGCHECKFLAG_GROUND)) {
+        this->actor.velocity.y = 13.0f;
+    }
+
+    if (!(curFrame < this->animEndFrame)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:839]: Whenever it comes to theses `!` conditions with floats, I wonder if that's an artificial result of early returns? i.e.
```c
if (curFrame < this->animEndFrame) {
    return;
}
... // rest of the code
```
Just speculation though

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 11]
```diff
@@ -81,106 +159,1546 @@ ActorInit En_Jso_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AF53C.s")
+Gfx* EnJso_SetAfterimageRenderMode(GraphicsContext* gfxCtx) {
+    Gfx* gfxHead = GRAPH_ALLOC(gfxCtx, 2 * sizeof(Gfx));
+    Gfx* gfx = gfxHead;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Jso/func_809AF5F8.s")
+    gDPSetRenderMode(
+        gfx++, AA_EN | Z_CMP | Z_UPD | IM_RD | CLR_ON_CVG | CVG_DST_WRAP | ZMODE_XLU | FORCE_BL | G_RM_FOG_SHADE_A,
+        AA_EN | Z_CMP | Z_UPD | IM_RD | CLR_ON_CVG | CVG_DST_WRAP | ZMODE_XLU | FORCE_BL |
+            GBL_c2(G_BL_CLR_IN, G_BL_A_IN, G_BL_CLR_MEM, G_BL_1MA));
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1627]: Why do I feel there's probably a macro that could simplify some of these args? But I don't know enough about gfx.h, maybe @Thar0 would know?

**Review by engineer124 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 10 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c]: Every single one matches like this, yeah. I pushed it up in this commit so you can see this change in isolation: https://github.com/zeldaret/mm/pull/1440/commits/464406e817046c173df60aa9f2aa707078797f53

What do you think? If you don't like it, I can just revert, it's no big deal.

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 11 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1641]: I should note that this function is a literal copy of `func_800BD9A0` in `z_actor.c`. So it has the same giant number of args there too. On the one hand, we can get two birds with one stone if we figure this out here. On the other hand, perhaps we already tried and failed to simplify this.

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 9 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:556]: I haven't gotten around to decompiling Garo Master yet (been too busy documenting this actor and Odolwa), but a quick look at the ASM shows it's not in there. And that makes sense, because the Garo Master doesn't use a ring of fire like regular Garos do.

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c [hunk 12]
```diff
@@ -81,106 +159,1560 @@ ActorInit En_Jso_InitVars = {
...
+
+    if ((curFrame >= this->animEndFrame) && (this->timer == 0)) {
+        EnJso_SetupFallDownAndTalk(this, play);
+    }
+}
+
+void EnJso_SetupFallDownAndTalk(EnJso* this, PlayState* play) {
+    EnJso_ChangeAnim(this, EN_JSO_ANIM_FALL_DOWN);
+    this->textIndex = 2;
+    this->actor.colChkInfo.mass = 255;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Jso/z_en_jso.c:1158]:
```suggestion
    this->actor.colChkInfo.mass = MASS_IMMOVABLE;
```

**Review by AngheloAlf — APPROVED** [review]
small nit

File: src/overlays/actors/ovl_En_Jso/z_en_jso.c (hunk 12 above)
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Jso/z_en_jso.c:1158]: Whoops, thought I caught all these, thanks.

**Review by hensldm — APPROVED** [review]
